### ISSUE TITLE
skb_push causes panic on skb without linear part

### ISSUE THREAD

**Issue by milabs** [tempesta-tech/tempesta#469]
When running Tempesta in QEMU with CPUx8/4G/virtio_net the BUG happens:

```
      KERNEL: vmlinux                           
    DUMPFILE: /home/ilya/vms/shared/dump2
        CPUS: 8
        DATE: Wed May  4 22:48:20 2016
      UPTIME: 00:11:34
LOAD AVERAGE: 0.25, 0.08, 0.03
       TASKS: 110
    NODENAME: tempesta
     RELEASE: 4.1.23-tfw
     VERSION: #68 SMP PREEMPT Wed May 4 22:16:44 MSK 2016
     MACHINE: x86_64  (4007 Mhz)
      MEMORY: 4 GB
       PANIC: "kernel BUG at net/core/skbuff.c:107!"
         PID: 0
     COMMAND: "swapper/1"
        TASK: ffff88015b084010  (1 of 8)  [THREAD_INFO: ffff88015b0d4000]
         CPU: 1
       STATE: TASK_RUNNING (PANIC)

crash> bt
PID: 0      TASK: ffff88015b084010  CPU: 1   COMMAND: "swapper/1"
 #0 [ffff88017fc439a0] machine_kexec at ffffffff8102ffad
 #1 [ffff88017fc439f0] crash_kexec at ffffffff810a25be
 #2 [ffff88017fc43ac0] oops_end at ffffffff810069f0
 #3 [ffff88017fc43af0] die at ffffffff81006b86
 #4 [ffff88017fc43b20] do_trap at ffffffff81003d6b
 #5 [ffff88017fc43b80] do_error_trap at ffffffff81003ed5
 #6 [ffff88017fc43c40] do_invalid_op at ffffffff8100442b
 #7 [ffff88017fc43c50] invalid_op at ffffffff81422b1b
    [exception RIP: skb_panic+0x5e]
    RIP: ffffffff813450ae  RSP: ffff88017fc43d08  RFLAGS: 00010296
    RAX: 0000000000000088  RBX: ffff880037bd2040  RCX: 0000000000000088
    RDX: 0000000080000103  RSI: ffffffff817552fd  RDI: 00000000ffffffff
    RBP: ffff88017fc43d28   R8: 0000000000000000   R9: 000000000000163c
    R10: 0000000000000001  R11: 000000000000163c  R12: ffff88014dde5400
    R13: 0000000000000020  R14: 0000000000000020  R15: 0000000000000000
    ORIG_RAX: ffffffffffffffff  CS: 0010  SS: 0018
 #8 [ffff88017fc43d30] skb_push at ffffffff8134590c
 #9 [ffff88017fc43d40] tcp_transmit_skb at ffffffff8139f889
#10 [ffff88017fc43dc0] tcp_write_xmit at ffffffff813a0373
#11 [ffff88017fc43e40] __tcp_push_pending_frames at ffffffff813a126d
#12 [ffff88017fc43e60] tcp_push at ffffffff8138f720
#13 [ffff88017fc43e70] ss_tx_action at ffffffffa004a111 [tempesta_fw]
#14 [ffff88017fc43ef0] net_tx_action at ffffffff81354993
#15 [ffff88017fc43f20] __do_softirq at ffffffff81048115
#16 [ffff88017fc43f80] irq_exit at ffffffff8104846d
#17 [ffff88017fc43f90] smp_trace_call_function_single_interrupt at ffffffff8102832e
#18 [ffff88017fc43fa0] smp_call_function_single_interrupt at ffffffff81028339
#19 [ffff88017fc43fb0] call_function_single_interrupt at ffffffff8142286b
--- <IRQ stack> ---
#20 [ffff88015b0d7e08] call_function_single_interrupt at ffffffff8142286b
    [exception RIP: native_safe_halt+0x6]
    RIP: ffffffff810340b6  RSP: ffff88015b0d7eb8  RFLAGS: 00000202
    RAX: 0000000000000001  RBX: ffffffff81068fe0  RCX: 0140000000000000
    RDX: 0100000000000000  RSI: ffffffff817368f5  RDI: ffffffff81740807
    RBP: ffff88015b0d7eb8   R8: 0000000000000000   R9: 0000000000000000
    R10: ffff88017fc541e8  R11: 0000000000000000  R12: ffff88015b0d4000
    R13: 000000a1a3bb19c6  R14: 0000000000000001  R15: ffff88015b0d4000
    ORIG_RAX: ffffffffffffff04  CS: 0010  SS: 0018
#21 [ffff88015b0d7ec0] default_idle at ffffffff8100c43e
#22 [ffff88015b0d7ed0] arch_cpu_idle at ffffffff8100cf3a
#23 [ffff88015b0d7ee0] cpu_startup_entry at ffffffff810747ff
#24 [ffff88015b0d7f30] start_secondary at ffffffff8102892b
```

The BUG and is reproducible and the reason is show below:
```
1421 unsigned char *skb_push(struct sk_buff *skb, unsigned int len)
1422 {
1423         skb->data -= len;
1424         skb->len  += len;
1425         if (unlikely(skb->data<skb->head))
1426                 skb_under_panic(skb, len, __builtin_return_address(0));
1427         return skb->data;
1428 }
1429 EXPORT_SYMBOL(skb_push);
```

The condition `if (unlikely(skb->data<skb->head))` is met and `skb_under_panic` triggers. Looking through the Tempesta's code leads me to strange [allocation](https://github.com/natsys/tempesta/blob/52faa3d1280dbae18cd52601765c3efc5abe915a/tempesta_fw/ss_skb.c#L187) of skb with 0 size:

```
		/* Allocate a new SKB to hold @n_excess page fragments. */
		nskb = alloc_skb(0, GFP_ATOMIC);
		if (nskb == NULL)
			return -ENOMEM;
```

Changing 0 to 256 helps and the BUG doesn't appear any more.

Here is the test I've run to catch the BUG and ensure that the BUG is gone:
ab -k -r -n 10000000 -c 1000 http://192.168.88.220:3128/index.html

**Comment by keshonok** [comment]
The SKBs allocated in `__extend_pgfrags()` are attached to the parent skb as skb fragments through `frag_list`. SKB fragments don't need the linear part. For this bug to happen, one of these SKBs must be turned into a standalone SKB in the egress path in the kernel.

**Comment by milabs** [comment]
Seems, that 178e09ded2e086fda810f9451c5c8057ca263bec fixes the issue.

**Comment by keshonok** [comment]
This and similar symptoms have appeared again as other parts of the code are developed. I believe that the patch mentioned above is unrelated to the issue.

Reopening the issue.

**Comment by milabs** [comment]
The problem has gone with removing the following code:

~~~
        nskb = alloc_skb(0, GFP_ATOMIC);
        if (nskb == NULL)
            return -ENOMEM;
~~~

Now, all the code uses `ss_skb_alloc` and it calls `alloc_skb(MAX_TCP_HEADER, GFP_ATOMIC);`. So, the kernel always have some space to push headers.

**Comment by keshonok** [comment]
This issue had been observed when one most notable bug had not been fixed yet. The bug was fixed by commit 8a569b2.

The fix eliminated the use of skb fragments in egress SKBs (those that are attached via `frag_list`). The presence of skb fragments caused the following major issue. SKBs are split in the egress path according to the interface's MTU and MSS. However, the kernel's `tcp_fragment()` function doesn't handle skb fragments, and simply discards them completely. At the same time, the lengths of those skb fragments are still taken into account in `skb->len` and `skb->datalen`. That leads to all kinds of strange combinations of `skb->len` and `skb->datalen`, and I wouldn't exclude a possibility of insufficient space if the linear data which is the difference between `skb->datalen` and `skb->len`.

#503 was the other notable bug that was fixed. It may have had some influence in combination with the bug mentioned above.

This issue has not been observed since these two bugs had been fixed.
